File: C++/1261.cpp
__________________________________________________________________________________________________
sample 20 ms submission
/*
Problem Statement: https://leetcode.com/problems/find-elements-in-a-contaminated-binary-tree/
*/

static const int __ = []() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);
    std::cout.tie(nullptr);
    return 0;
}();

class FindElements {
private:
	unordered_set<int> seen;
public:
    FindElements(TreeNode* root) {
		fix(0, root);
    }

	void fix(int val, TreeNode* node) {
		seen.insert(val);
		if (node->left)
			fix(2 * val + 1, node->left);
		if (node->right)
			fix(2 * val + 2, node->right);
	}
    
    bool find(int target) {
		return seen.find(target) != seen.end();
    }
};
__________________________________________________________________________________________________
sample 24 ms submission
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class FindElements {
public:
    unordered_set<int> set;
    FindElements(TreeNode* root) {
        recover(root, 0);
    }
    
    void recover(TreeNode* root, int x){
        if(root==NULL) return;
        root->val = x;
        set.emplace(x);
        if(root->left) recover(root->left,  2 * x + 1);
        
        if(root->right) recover(root->right,  2 *x + 2);
        
        
    }
    
    bool find(int target) {
        return set.count(target);
        
    }
};

/**
 * Your FindElements object will be instantiated and called as such:
 * FindElements* obj = new FindElements(root);
 * bool param_1 = obj->find(target);
 */
__________________________________________________________________________________________________

File: Python3/1261.py
__________________________________________________________________________________________________
sample 64 ms submission
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class FindElements:
    def __init__(self, root: TreeNode):
        self.root = root
        self.d = {}
        if self.root is not None:
            self.root.val = 0
            self.fillin(self.root)
            self.d[0] = True
            
    def fillin(self, root):
        if root is None or (root.left is None and root.right is None):
            return
        if root.left is not None and root.right is not None:
            root.left.val = root.val * 2 + 1
            root.right.val = root.val * 2 + 2
            self.d[root.left.val] = True
            self.d[root.right.val] = True
            self.fillin(root.left)
            self.fillin(root.right)
        elif root.left is not None:
            root.left.val = root.val * 2 + 1
            self.d[root.left.val] = True
            self.fillin(root.left)
        else:
            root.right.val = root.val * 2 + 2
            self.d[root.right.val] = True
            self.fillin(root.right)

    def find(self, target: int) -> bool:
        if not self.d:
            return False
        return target in self.d

# Your FindElements object will be instantiated and called as such:
# obj = FindElements(root)
# param_1 = obj.find(target)
__________________________________________________________________________________________________
sample 68 ms submission
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class FindElements:
    
    def recoverTree(self, root, val):
        
        if not root:
            return
        
        root.val = val
        self.values.add(val)
        if root.left:
            self.recoverTree(root.left, 2 * val + 1)
        if root.right:
            self.recoverTree(root.right, 2 * val + 2)

    def __init__(self, root: TreeNode):
        self.values = set()
        self.recoverTree(root, 0)
__________________________________________________________________________________________________
